torch/csrc/jit/passes/annotate_effects.cpp

```diff
@@ -0,0 +1,227 @@
+      auto falseBlock = node->blocks().at(1);
+
+      auto trueToken = visitBlock(trueBlock, curToken);
+      auto falseToken = visitBlock(falseBlock, curToken);
+
+      // If any branch has a mutating op, this node has to output a world token
+      if (trueToken != curToken || falseToken != curToken) {
+        trueBlock->registerOutput(trueToken);
+        falseBlock->registerOutput(falseToken);
+
+        return node->addOutput()->setType(WorldType::get());
+      }
+      return curToken;
+    }
+
+    if (node->kind() == prim::Loop) {
+      JIT_ASSERT(node->blocks().size() == 1);
+      auto block = node->blocks().at(0);
+      if (!shouldAnnotate(block)) {
+        // Bail out early if there's no mutable variables used inside
+        return curToken;
+      }
+
+      // Register the world token as a loop carried dependency
+      auto beginLoopToken = block->addInput()->setType(WorldType::get());
+      auto endLoopToken = visitBlock(block, beginLoopToken);
+      block->registerOutput(endLoopToken);
+
+      JIT_ASSERT(endLoopToken != beginLoopToken);
+
+      // Thread the world token through the loop node
+      node->addInput(curToken);
+      return node->addOutput()->setType(WorldType::get());
+    }
+
+    // For mutating ops, just thread the world token through the node.
+    if (isMutatingOp(node)) {
+      node->replaceInput(0, curToken);
+      return node->outputs().at(0);
+    }
+
+    JIT_ASSERT(node->blocks().size() == 0);
+
+    // For pure ops that need to be annotated, fence them.
+    if (shouldAnnotate(node)) {
+      return addFenceForNode(node, curToken);
+    }
+
+    return curToken;
+  }
+
+  bool shouldAnnotate(Node* node) {
+    // Check if this node uses a known mutable value
+    for (auto* input : node->inputs()) {
+      if (input->type()->kind() != TypeKind::ListType) {
+        // TODO(suo): Right now, we only support mutable lists.
+        // If we remove this check, it's not clear whether:
+        //
+        //   append(int[] a, int b)
+        //
+        // mutates `a` or `b`. We'll need to extend the schema language to be
+        // able to express which argument is mutated.
+        continue;
+      }
+      // First check the cache
+      if (mutableValues_.count(input) != 0) {
+        return true;
+      }
+
+      // Check whether any mutating op uses this input
+      for (const auto& use : input->uses()) {
+        if (isMutatingOp(use.user)) {
+          mutableValues_.insert(input);
+          return true;
+        }
+      }
+    }
+
+    // Check that any sub-blocks need to be annotated
+    for (auto block : node->blocks()) {
+      if (shouldAnnotate(block)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  bool shouldAnnotate(Block* block) {
+    for (auto node : block->nodes()) {
+      if (shouldAnnotate(node)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  bool isMutatingOp(const Node* node) {
+    return std::any_of(
```

Comment:
Shouldn't this be: `return node->inputs().size() > 0 && node->inputs()[0]->type() == WorldType::get()`. When you replace the world node you only ever replace the first entry.